File: src/core/ntsp.cpp
#include "ntsp.h"
#include <fstream>
#include <map>
#include <memory>
#include <unordered_map>

namespace sf {

static uint32_t name_hash(const std::string& s) {
    uint32_t h = 0;
    for (unsigned char c : s) h = h * 31 + c;
    return h & 0x7FFFFFFF;
}

// Cached read-only PSTN package: entry hash -> (blobIndex, blobCount), + blob table.
struct Package {
    std::string path;
    std::ifstream f;
    std::unordered_map<uint32_t, std::pair<uint32_t, uint32_t>> by_hash;  // hash -> (blobIndex,count)
    Bytes head;
    size_t blob_base = 0;
    bool ok = false;

    explicit Package(const std::string& p) : path(p), f(p, std::ios::binary) {
        if (!f) return;
        uint8_t hdr[24];
        f.read((char*)hdr, 24);
        if (memcmp(hdr, "PSTN", 4) != 0) return;
        uint32_t entry_count = u32le(hdr, 8), blob_count = u32le(hdr, 12);
        uint64_t header_size = u64le(hdr, 16);
        head.resize(header_size);
        f.seekg(0);
        f.read((char*)head.data(), header_size);
        blob_base = 0x18 + (size_t)entry_count * 0x18;
        (void)blob_count;
        for (uint32_t i = 0; i < entry_count; i++) {
            size_t o = 0x18 + (size_t)i * 0x18;
            by_hash[u32le(head.data(), o)] = {u32le(head.data(), o + 4), u32le(head.data(), o + 8)};
        }
        ok = true;
    }

    bool mips(const std::string& name, Bytes& out) {
        auto it = by_hash.find(name_hash(name));
        if (it == by_hash.end()) return false;
        uint32_t bi = it->second.first, bc = it->second.second;
        for (uint32_t k = 0; k < bc; k++) {
            size_t o = blob_base + (size_t)(bi + k) * 0x10;
            if (o + 16 > head.size()) return false;
            uint64_t off = u64le(head.data(), o), sz = u64le(head.data(), o + 8);
            size_t cur = out.size();
            out.resize(cur + sz);
            f.seekg(off);
            f.read((char*)out.data() + cur, sz);
        }
        return true;
    }
};

static std::map<std::string, std::unique_ptr<Package>>& cache() {
    static std::map<std::string, std::unique_ptr<Package>> c;
    return c;
}

bool resolve_streamed_dds(const Bytes& stub, const std::string& tex_base,
                          const std::string& streaming_dir, Bytes& out) {
    if (!is_ntsi(stub) || stub.size() < 0x18) return false;
    const uint8_t* d = stub.data();
    uint32_t pns = u32le(d, 0x0C), m4s = u32le(d, 0x10);
    std::string pkg_name((const char*)d + 0x18);   // nul-terminated within packageNameSize
    if (pkg_name.size() > pns) pkg_name.resize(pns);
    size_t hdr_off = 0x18 + pns + m4s;
    if (hdr_off + 0x80 > stub.size()) return false;
    bool dx10 = memcmp(d + hdr_off + 0x54, "DX10", 4) == 0;
    size_t hdr_size = dx10 ? 0x94 : 0x80;

    std::string pkg_path = streaming_dir + "\\" + pkg_name + ".ntsp";
    auto& c = cache();
    auto it = c.find(pkg_path);
    if (it == c.end())
        it = c.emplace(pkg_path, std::make_unique<Package>(pkg_path)).first;
    if (!it->second->ok) return false;

    out.assign(d + hdr_off, d + hdr_off + hdr_size);   // DDS header from the stub
    return it->second->mips(tex_base, out);            // append streamed mip data
}

} // namespace sf

File: tools/ntsp.py
"""Needle Texture Streaming (NTSI stub in .pac -> PSTN .ntsp package) resolver.

A Frontiers .dds inside a .pac may be an NTSI stub (real pixels streamed). This
reconstructs the full .dds: read the stub's DDS header + package name, look the
texture up in <package_name>.ntsp by name hash, concatenate its mip blobs.
Format per HedgeLib++ hl_hh_needle_texture_streaming.h.
"""
import os, struct


def compute_name_hash(name: str) -> int:
    h = 0
    for c in name.encode("utf-8"):
        h = (h * 31 + c) & 0xFFFFFFFF
    return h & 0x7FFFFFFF


def is_ntsi(data: bytes) -> bool:
    return len(data) >= 4 and data[:4] == b"NTSI"


def parse_ntsi(data: bytes):
    """Return dict with package_name and dds_header (bytes)."""
    sig, ver, unk0, pns, m4s, m4i = struct.unpack_from("<6I", data, 0)
    assert sig == 0x4953544E  # "NTSI"
    name = data[0x18:0x18 + pns].split(b"\x00")[0].decode("utf-8", "replace")
    hdr_off = 0x18 + pns + m4s
    # header size: 0x94 if DX10 fourcc else 0x80
    fourcc = data[hdr_off + 0x54:hdr_off + 0x58]
    hdr_size = 0x94 if fourcc == b"DX10" else 0x80
    return {
        "package_name": name,
        "dds_header": data[hdr_off:hdr_off + hdr_size],
        "mip4x4": data[0x18 + pns:0x18 + pns + m4s],
        "mip4x4_index": m4i,
    }


class Package:
    """Read-only PSTN package. Indexes entries by name hash; loads blob data on demand."""
    def __init__(self, path):
        self.path = path
        with open(path, "rb") as f:
            sig, ver, self.entry_count, self.blob_count = struct.unpack("<4I", f.read(16))
            self.header_size = struct.unpack("<Q", f.read(8))[0]
            f.seek(0)
            head = f.read(self.header_size)
        assert sig == 0x4E545350  # "PSTN"
        self.head = head
        self.blob_base = 0x18 + self.entry_count * 0x18
        # index entries by hash
        self.by_hash = {}
        for i in range(self.entry_count):
            o = 0x18 + i * 0x18
            nh, bi, bc = struct.unpack_from("<3I", head, o)
            w, h = struct.unpack_from("<2H", head, o + 12)
            self.by_hash[nh] = (bi, bc, w, h)

    def blob(self, index):
        o = self.blob_base + index * 0x10
        data_off, data_size = struct.unpack_from("<QQ", self.head, o)
        with open(self.path, "rb") as f:
            f.seek(data_off)
            return f.read(data_size)

    def get_mips(self, name):
        """Return concatenated mip bytes for a texture name, or None."""
        e = self.by_hash.get(compute_name_hash(name))
        if not e:
            return None, None
        bi, bc, w, h = e
        return b"".join(self.blob(bi + k) for k in range(bc)), (w, h)


_pkg_cache = {}


def resolve_dds(stub_bytes: bytes, texture_name: str, streaming_dir: str):
    """Reconstruct a full .dds from an NTSI stub. texture_name = dds base (no ext).
    Returns full .dds bytes, or the stub unchanged if not NTSI / package missing."""
    if not is_ntsi(stub_bytes):
        return stub_bytes
    info = parse_ntsi(stub_bytes)
    pkg_path = os.path.join(streaming_dir, info["package_name"] + ".ntsp")
    if not os.path.exists(pkg_path):
        return None
    pkg = _pkg_cache.get(pkg_path)
    if pkg is None:
        pkg = _pkg_cache[pkg_path] = Package(pkg_path)
    mips, dims = pkg.get_mips(texture_name)
    if mips is None:
        return None
    return info["dds_header"] + mips


if __name__ == "__main__":
    import sys
    from binlib import hexdump
    import pac
    ents = pac.unpack(sys.argv[1])                       # a .pac
    name = sys.argv[2]                                   # e.g. isl_obj_portalbit_body_abd.dds
    streaming = r"E:\Games\steamapps\common\SonicFrontiers\image\x64\raw\texture_streaming"
    stub = [e for e in ents if e.name == name][0].data
    base = name[:-4] if name.endswith(".dds") else name
    dds = resolve_dds(stub, base, streaming)
    print(f"stub {len(stub)}B -> full dds {len(dds) if dds else None}B")
    if dds:
        w = struct.unpack_from("<I", dds, 0x10)[0]; h = struct.unpack_from("<I", dds, 0x0C)[0]
        print(f"dds {w}x{h} fourcc={dds[0x54:0x58]}  mip0 expected DXT1={w*h//2}")
        print(hexdump(dds, 0, 32))
